core/src/main/java/hudson/model/ParametersAction.java

```diff
@@ -326,7 +326,7 @@ public void onLoad(Run<?, ?> r) {
         for (ParameterValue v : this.parameters) {
             if (this.parameterDefinitionNames.contains(v.getName()) || isSafeParameter(v.getName())) {
                 filteredParameters.add(v);
-            } else if ("false".equalsIgnoreCase(shouldKeepFlag)) {
+            } else if (!"false".equalsIgnoreCase(shouldKeepFlag)) {
```

Comment:
I updated to use optBoolean, which should handle the true, false and null cases as intended.